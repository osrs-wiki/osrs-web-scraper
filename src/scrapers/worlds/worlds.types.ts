export type World = {
  activity: string;
  number: number;
  region: string;
  type: "deadman" | "yes" | "no";
};
