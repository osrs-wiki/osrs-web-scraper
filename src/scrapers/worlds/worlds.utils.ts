import { MediaWikiTemplate } from "@osrs-wiki/mediawiki-builder";
import { HTMLElement } from "node-html-parser";

export const WORLD_LIST_URL = "https://oldschool.runescape.com/a=13/slu";

export const getWorldLines = (worldRows: HTMLElement) => {
  const worldRowNodes = worldRows.childNodes.filter(
    (node) => node instanceof HTMLElement && node.tagName === "TR"
  );
  const worldLines = worldRowNodes.map((node) => {
    const tdNodes = node.childNodes.filter(
      (node) => node instanceof HTMLElement && node.tagName === "TD"
    );
    const worldLine = new MediaWikiTemplate("WorldLine", { collapsed: true });
    const worldNumber =
      tdNodes[0].childNodes?.[1].textContent?.replaceAll(/^\D+/g, "") ?? "";
    const region = tdNodes[2].textContent;
    const activity = tdNodes[4].textContent;
    const members = activity.includes("Deadman")
      ? "deadman"
      : tdNodes[3].textContent === "Members"
      ? "yes"
      : "no";
    worldLine.add("", worldNumber);
    worldLine.add("", region);
    worldLine.add("mems", members);
    worldLine.add("", activity);
    return worldLine;
  });
  return worldLines;
};
