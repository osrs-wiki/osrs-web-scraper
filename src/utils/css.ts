import { HTMLElement } from "node-html-parser";

/**
 * Extracts background image URL from CSS style attribute
 * Handles both single and double quotes, and HTML entity encoding
 */
export function extractBackgroundImageUrl(style: string): string | null {
  if (!style) return null;
  
  // First decode HTML entities
  const decodedStyle = style
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&');
  
  // Match background-image: url("...") or url('...') patterns
  const urlMatch = decodedStyle.match(/background-image\s*:\s*url\s*\(\s*['"']?([^'"')]+)['"']?\s*\)/i);
  
  if (urlMatch && urlMatch[1]) {
    return urlMatch[1];
  }
  
  return null;
}

/**
 * Extracts all background image URLs from style attributes in HTML elements
 */
export function extractBackgroundImages(elements: HTMLElement[]): string[] {
  const imageUrls: string[] = [];
  
  elements.forEach(element => {
    if (element.attributes?.style) {
      const url = extractBackgroundImageUrl(element.attributes.style);
      if (url) {
        imageUrls.push(url);
      }
    }
  });
  
  return imageUrls;
}