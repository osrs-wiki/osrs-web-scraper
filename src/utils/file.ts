import fs from "fs";
import client from "https";
import path from "path";

import { formatText } from "./text";

/**
 * Download a file from a url
 * @param url The url of the file to download
 * @param filepath The filepath of the downloaded file
 * @returns
 */
export const downloadFile = async (url: string, filepath: string) => {
  console.info(`Attempting file download: ${url}`);
  return new Promise((resolve, reject) => {
    client.get(url, (res) => {
      if (res.statusCode === 200) {
        res
          .pipe(fs.createWriteStream(filepath))
          .on("error", reject)
          .once("close", () => resolve(filepath));
        console.info(`Downloaded file ${filepath}`);
      } else {
        res.resume();
        reject(
          new Error(`Request Failed With a Status Code: ${res.statusCode}`)
        );
      }
    });
  });
};

/**
 * Get a file from a url
 * @param fileUrl The file url
 * @returns
 */
export const getFileName = (fileUrl: string) => {
  const parsed = new URL(fileUrl);
  return path.basename(parsed.pathname);
};

/**
 * Get the extension of a file from a url link.
 * @param fileUrl The file url
 * @returns
 */
export const getFileExtension = (fileUrl: string) => {
  return fileUrl.split(/[#?]/)[0].split(".").pop().trim();
};

const fileCharsReplaceMap: { [key: string]: string } = {
  ":": "-",
  "|": "-",
  "&nbsp;": "",
};

/**
 * Format a file name to ensure it is valid.
 * @param fileName The file name
 * @returns
 */
export const formatFileName = (fileName: string) => {
  const cleanedFileName = Object.keys(fileCharsReplaceMap).reduce(
    (name, find) => name.replaceAll(find, fileCharsReplaceMap[find]),
    fileName
  );
  return formatText(cleanedFileName);
};
