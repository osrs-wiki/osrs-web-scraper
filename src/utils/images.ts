import fs from "fs";
import client from "https";
import path from "path";

import { formatText } from "./text";

/**
 * Download an image from a url
 * @param url The url of the image to download
 * @param filepath The filepath of the downloaded image
 * @returns
 */
export const downloadImage = async (url: string, filepath: string) => {
  console.info(`Attempting image download: ${url}`);
  return new Promise((resolve, reject) => {
    client.get(url, (res) => {
      if (res.statusCode === 200) {
        res
          .pipe(fs.createWriteStream(filepath))
          .on("error", reject)
          .once("close", () => resolve(filepath));
        console.info(`Downloaded image ${filepath}`);
      } else {
        res.resume();
        reject(
          new Error(`Request Failed With a Status Code: ${res.statusCode}`)
        );
      }
    });
  });
};

/**
 * Get an image name from a url
 * @param imageUrl The image url
 * @returns
 */
export const getImageName = (imageUrl: string) => {
  const parsed = new URL(imageUrl);
  return path.basename(parsed.pathname);
};

/**
 * Get the extension of an image from a url link.
 * @param imageUrl The image url
 * @returns
 */
export const getImageExtension = (imageUrl: string) => {
  return imageUrl.split(/[#?]/)[0].split(".").pop().trim();
};

/**
 * Format a file name to ensure it is valid.
 * @param fileName The file name
 * @returns
 */
export const formatFileName = (fileName: string) => {
  return formatText(fileName.replaceAll(":", "-"));
};
