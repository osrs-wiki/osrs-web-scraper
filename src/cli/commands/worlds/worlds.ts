import { Command } from "commander";

import scraper from "../../../scraper";
import { worlds as worldsScraper } from "../../../scrapers";
import { WORLD_LIST_URL } from "../../../scrapers/worlds/worlds.utils";

const worlds = new Command("worlds")
  .description("Scrape the OSRS world list.")
  .action(() => {
    scraper.scrape(WORLD_LIST_URL, worldsScraper);
  });

export default worlds;
