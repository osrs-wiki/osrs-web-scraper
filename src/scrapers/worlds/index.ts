import worlds from "./worlds";

export default worlds;
