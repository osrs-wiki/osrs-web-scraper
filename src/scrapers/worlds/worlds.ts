import { MediaWikiBuilder } from "@osrs-wiki/mediawiki-builder";
import fs from "fs";
import parse from "node-html-parser";

import { getWorldLines } from "./worlds.utils";
import { ScrapingService } from "../types";

const worlds: ScrapingService<MediaWikiBuilder> = {
  scrape: async (page): Promise<MediaWikiBuilder> => {
    try {
      const results = await page.evaluate(() => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore Ignore window typing
        const $ = window.$;
        const worldRows = $(".server-list__body").html();

        return {
          worldRows,
        };
      });

      const worldNodes = parse(results.worldRows);

      const builder = new MediaWikiBuilder();
      builder.addContents(getWorldLines(worldNodes));

      console.info("Writing world list results to file...");
      try {
        if (!fs.existsSync("out/worlds")) {
          fs.mkdirSync("out/worlds", { recursive: true });
        }
        await fs.writeFileSync(`out/worlds/worlds.txt`, builder.build());
        console.info("Successfully created worlds file");
      } catch (err) {
        console.error(err);
      }

      return builder;
    } catch (error) {
      console.error(error);
    }
  },
};

export default worlds;
