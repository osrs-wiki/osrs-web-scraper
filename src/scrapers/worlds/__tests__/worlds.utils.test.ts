import { MediaWikiBuilder } from "@osrs-wiki/mediawiki-builder";
import parse from "node-html-parser";

import { getWorldLines } from "../worlds.utils";

describe("world scraper utils", () => {
  test("getWorldLines", () => {
    const worldNodes = parse(
      '<tr><td>\n<a href="">OldSchool 101</a></td><td>0 players</td><td>United States</td><td>Members</td><td>111-126 Deadman</td></tr>'
    );
    const builder = new MediaWikiBuilder();
    builder.addContents(getWorldLines(worldNodes));
    expect(builder.build()).toMatchSnapshot();
  });
});
