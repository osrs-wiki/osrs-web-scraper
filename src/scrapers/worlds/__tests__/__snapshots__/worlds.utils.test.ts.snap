// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`world scraper utils getWorldLines 1`] = `
"{{WorldLine|101|United States|mems=deadman|111-126 Deadman}}
"
`;
